Improve readme (#16) + temp fix for KeyError bug (#17) (#18)
bellingcat/wayback-google-analytics: README.md

        <li><a href="#using-poetry-for-development">Using Poetry for development</li>
      </ul>
    </li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->
## About The Project

Wayback Google Analytics is a lightweight tool that gathers current and historic
Google analytics data (UA, GA and GTM codes) from a collection of website urls.

### Why do I need GA codes?

Google Analytics codes are a useful data point when examining relationships between websites. If two seemingly disparate websites share the same UA, GA or GTM code then there is a good chance that they are managed by the same individual or group. This useful breadcrumb has been used by researchers and journalists in OSINT investigations regularly over the last decade, but a recent change in how Google handles its analytics codes threatens to limit its effectiveness. Google began phasing out UA codes as part of its Google Analytics 4 upgrade in July 2023, making it significantly more challenging to use this breadcrumb during investigations.

### How does this tool help me?

> Luckily, the Internet Archive's Wayback Machine contains useful snapshots of websites containing their historic GA IDs. While you could feasibly check each snapshot manually, this tool automates that process with the Wayback Machines CDX API to simplify and speed up the process. Enter a list of urls and a time frame (along with extra, optional parameters) to collect current and historic GA, UA and GTM codes and return them in a format you choose (json, txt, xlsx or csv).

Reviewer: ```suggestion
Luckily, the Internet Archive's [Wayback Machine](https://archive.org/web/) contains useful snapshots of websites containing their historic GA IDs. While you could feasibly check each snapshot manually, this tool automates that process with the Wayback Machines CDX API to simplify and speed up the process. Enter a list of urls and a time frame (along with extra, optional parameters) to collect current and historic GA, UA and GTM codes and return them in a format you choose (json, txt, xlsx or csv).
```
link to the wayback machine if people aren't familiar.

    }
```

### Further reading

> - For more info about analytics codes and what the GA-4 rollout means for OSINT: [https://digitalinvestigations.substack.com/p/what-the-rollout-of-google-analytics]

Reviewer: ```suggestion
- For more info about analytics codes and what the GA-4 rollout means for OSINT: https://digitalinvestigations.substack.com/p/what-the-rollout-of-google-analytics
```
not sure if the brackets were intentional or incomplete markdown, disregard if intentional.

> - For an example investigation usings analytics codes: [https://www.bellingcat.com/resources/how-tos/2015/07/23/unveiling-hidden-connections-with-google-analytics-ids/]

Reviewer: ```suggestion
- For an example investigation usings analytics codes: https://www.bellingcat.com/resources/how-tos/2015/07/23/unveiling-hidden-connections-with-google-analytics-ids/
```
not sure if the brackets were intentional or incomplete markdown, disregard if intentional.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

### Built With

<div align="center">

![Python][Python]
![Pandas][Pandas]

</div>

Additional libraries/tools: [BeautifulSoup4](https://pypi.org/project/beautifulsoup4/), [Asyncio](https://docs.python.org/3/library/asyncio.html), [Aiohttp](https://docs.aiohttp.org/en/stable/)

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- GETTING STARTED -->
## Installation

> ### Install with pip

Reviewer: ```suggestion
### Install from [pypi](https://pypi.org/project/wayback-google-analytics/) (with pip)

[![PyPI][pypi-shield]][pypi-url]
```

Hope I got the shield thing right with the links at the end, made 2 suggestions.

The easiest way to to install Wayback Google Analytics is from the command line with pip.

1. Open a terminal window and navigate to your chosen directory.

> 2. Create a virtual environment and activate it (optional, but recommended)

Reviewer: ```suggestion
2. Create a virtual environment and activate it (optional, but recommended; if you use [Poetry](https://python-poetry.org/) or [pipenv](https://pipenv.pypa.io/en/latest/) those package managers do it for you)
```

```terminal
wayback-google-analytics -u https://yapatriot.ru https://zanogu.com https://whoswho.com.ua https://adamants.ru -s 01/01/2015 -f yearly -o xlsx
```

The result is a single `.xlsx` file with two sheets.

Ordered by website:
<div align="center">
  <img src="./docs/imgs/xlsxbysite.png" width=1000>
</div>

Ordered by code:
<div align="center">
  <img src="./docs/imgs/xlsxbycode.png">
</div>

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- CONTRIBUTING -->
## Contributing

### Bugs and feature requests

Please feel free to [open an issue](https://github.com/bellingcat/wayback-google-analytics/issues) should you encounter any bugs or have suggestions for new features or improvements. You can also [reach out to me](#contact) directly with suggestions or thoughts.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- DEVELOPMENT -->
## Development

### Testing

* Run tests with `python -m unittest discover`
* Check coverage with `coverage run -m unittest`

### Using Poetry for Development

> Wayback Google Analytics uses [Poetry](https://python-poetry.org/), a Python dependency management and packaging tool. A github workflow automates the publishing process on pushes to main ((see our workflow here)[https://github.com/bellingcat/wayback-google-analytics/actions]), but be sure to update the version number in `pyproject.toml` when updating.

Reviewer: ```suggestion
Wayback Google Analytics uses [Poetry](https://python-poetry.org/), a Python dependency management and packaging tool. A GitHub workflow automates the tests on PRs and to main ([see our workflow here](https://github.com/bellingcat/wayback-google-analytics/actions)),  be sure to update the [semantic](https://semver.org/) version number in `pyproject.toml` when opening a PR. 
```

> To publish a new version to PyPI, you can follow these steps:

Reviewer: ```suggestion
If you have push access, follow these steps to trigger the GitHub workflow that will build and release a new version to PyPI :
```

1. Change the version number in [pyproject.toml](pyproject.toml)
2. Create a new tag for that version `git tag "vX.0.0"`
3. Push the tag `git push --tags`

<!-- LICENSE -->
## License

> Distributed under the MIT License. See `LICENSE.txt` for more information.

Reviewer: License file seems to be missing.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- CONTACT -->
## Contact

You can contact me through email or social media.

> * email: jclarksummit@gmail.com

Reviewer: ```suggestion
* email: jclarksummit at gmail dot com
```
not sure how publicly widespread this email address is, but GH is easy to scrape and can easily be used for automated phishing campaigns. do disregard if that's not a concern

* Twitter/X: [@JustinClarkJO](https://twitter.com/JustinClarkJO)
* Linkedin: [Justin Clark](https://linkedin.com/in/justin-w-clark)

Project Link: [https://github.com/bellingcat/wayback-google-analytics](https://github.com/bellingcat/wayback-google-analytics)

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- ACKNOWLEDGMENTS -->
## Acknowledgments

* [Bellingcat](https://bellingcat.org) for hosting this project

> * [Miguel Ramalho](https://github.com/msramalho) for constant support, thoughtful code reviews and suggesting the original idea for this project

Reviewer: :rainbow:

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/bellingcat/wayback-google-analytics.svg?style=for-the-badge
[contributors-url]: https://github.com/bellingcat/wayback-google-analytics/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/bellingcat/wayback-google-analytics.svg?style=for-the-badge
[forks-url]: https://github.com/bellingcat/wayback-google-analytics/network/members
[stars-shield]: https://img.shields.io/github/stars/bellingcat/wayback-google-analytics.svg?style=for-the-badge
[stars-url]: https://github.com/bellingcat/wayback-google-analytics/stargazers
[issues-shield]: https://img.shields.io/github/issues/bellingcat/wayback-google-analytics.svg?style=for-the-badge

> [issues-url]: https://github.com/bellingcat/wayback-google-analytics/issues

Reviewer: ```suggestion
[issues-url]: https://github.com/bellingcat/wayback-google-analytics/issues
[pypi-shield]: https://img.shields.io/pypi/v/wayback-google-analytics
[pypi-url]: https://pypi.org/project/wayback-google-analytics/
```
